src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,128 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= elements.length) {
+            T[] newElements = (T[]) new Object[elements.length * 2];
+            System.arraycopy(elements, 0, newElements, 0, elements.length);
+            elements = newElements;
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        if (size == elements.length) {
+            T[] newElements = (T[]) new Object[elements.length * 2];
+            System.arraycopy(elements, 0, newElements, 0, elements.length);
+            elements = newElements;
+        }
+        for (int i = size - 1; i >= index; i--) {
+            elements[i + 1] = elements[i];
+        }
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int listSize = list.size();
+        if (size + listSize > elements.length) {
+            T[] newElements = (T[]) new Object[size + listSize];
+            for (int i = 0; i < size; i++) {
+                newElements[i] = elements[i];
+            }
+            elements = newElements;
+        }
 
+        for (int i = 0; i < listSize; i++) {
+            if (size + i >= elements.length) {
+                throw new ArrayListIndexOutOfBoundsException("Invalid array range: " + list);
+            }
+            elements[size + i] = list.get(i);
+        }
+        size += listSize;
```

Comment:
dont invent problems and difficulties in your code :)
```suggestion
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```